Add PortraitNet.

txyugood opened the pull request:
1. Add portraitNet model.
2. Add RandomAffine in transform.py for data augmentation.
3. Add kl_loss and focal_loss for PortraintNet.
4. Modify train.py for PortraintNet.
5. Add EG1800 and Supervisely dataset.
6. Add stepdecay in config.py for training strategy.

txyugood committed:
Add PortraitNet.

LutaoChu commented on paddleseg/transforms/transforms.py:
```diff
@@ -932,3 +910,87 @@ def __call__(self, im, label=None):
+
+
+@manager.TRANSFORMS.add_component
+class RandomAffine:
```
这个指缩放+旋转吧？是否可以通过已有的StepScaling+Rotation组合就好了，不用另外开发？

txyugood replied:
这个是平移、缩放和旋转。我之前试过通过StepScaling+Rotation但是效果完全不一样。所以按照论文实现了RandomAffine。

LutaoChu commented on paddleseg/datasets/eg1800.py:
```diff
@@ -0,0 +1,208 @@
+import copy
+
+
+def padding(img_ori, mask_ori, size=224, padding_color=128):
```
这些transform之前不是说可以复用的复用原来的transform，没有的新增到transform.py中吗？

txyugood replied:
这个padding跟transform中的padding略微有点区别，这个是居中的，transform里面的是是靠左的，而且不能设置。那我修改一下transform里的padding?

LutaoChu commented on paddleseg/datasets/supervisely.py:
```diff
@@ -0,0 +1,62 @@
+
+
+@manager.DATASETS.add_component
+class SUPERVISELY(EG1800):
```
尽量继承DataSet吧，数据集间解耦

txyugood replied:
好的，我改下这个。

LutaoChu commented on paddleseg/datasets/eg1800.py:
```diff
@@ -0,0 +1,208 @@
+    return diff
+    '''
+    cv2.drawContours(myImg, countours, -1, 1, 4)
+    # myImg[myImg==1] = 255
```
删掉注释

LutaoChu commented on paddleseg/models/portraitnet.py:
```diff
@@ -0,0 +1,389 @@
+
+
+@manager.MODELS.add_component
+class MobileNetV2(nn.Layer):
```
准确的讲，mobilenetv2是作为portraitnet的backbone吧
建议单独放一个文件，和mobilnetv3一样，放到paddleseg/models/backbone下

LutaoChu commented on paddleseg/models/portraitnet.py:
```diff
@@ -0,0 +1,389 @@
+        self.stage7 = InvertedResidual(self.depth(160), self.depth(320), 1,
+                                       6)  # 1/32
+
+        self.deconv1 = nn.Conv2DTranspose(self.depth(96),
```
decode的部分放到portraitnet中

LutaoChu commented on paddleseg/models/portraitnet.py:
```diff
@@ -0,0 +1,389 @@
+
+
+@manager.MODELS.add_component
+class PortraitNet(nn.Layer):
```
这段放在文件的前面吧

txyugood committed:
Add some transforms for PortraitNet.

txyugood committed:
Merge branch 'develop' into dev

LutaoChu commented on paddleseg/core/train.py:
```diff
@@ -104,8 +109,10 @@ def train(model,
         ddp_model = paddle.distributed.fleet.distributed_model(model)
 
-    batch_sampler = paddle.io.DistributedBatchSampler(
-        train_dataset, batch_size=batch_size, shuffle=True, drop_last=True)
```
yapf版本改为0.26.0，所有文件重新格式化一下吧

LutaoChu commented on paddleseg/datasets/eg1800.py:
```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
所有文件license里的2020改成2021

LutaoChu commented on paddleseg/datasets/eg1800.py:
```diff
@@ -0,0 +1,119 @@
+import paddleseg.transforms.functional as F
+import cv2
+import numpy as np
+import copy
```
所有文件的import顺序应该符合PEP8规范

txyugood committed:
Format code of PortraitNet.

txyugood committed:
remove unused code in eg1800.py.

LutaoChu committed:
Update portraitnet_eg8000_224x224_46k.yml

LutaoChu commented on configs/portraitnet/portraitnet_eg8000_224x224_46k.yml:
```diff
@@ -0,0 +1,76 @@
+  type: adam
+  weight_decay: 5.0e-4
+
+learning_rate:
```
最新用lr_scheduler，learning_rate要被废弃了

LutaoChu commented on paddleseg/datasets/eg1800.py:
```diff
@@ -0,0 +1,117 @@
+    EG1800 dataset `http://xiaoyongshen.me/webpage_portrait/index.html`.
+
+    Args:
+        transforms (list): A list of image transformations.
```
缺少transforms2的注释

LutaoChu commented on paddleseg/datasets/supervisely.py:
```diff
@@ -0,0 +1,116 @@
+            # add augmentation
+            std = np.array(self.std)[:, np.newaxis, np.newaxis]
+            mean = np.array(self.mean)[:, np.newaxis, np.newaxis]
+            im_aug = im * std + mean
```
这个地方逻辑有点别扭。最好是经过共同的transform后，出来两个分支transform，分别处理im和im_aug。
EG1800也是一样

LutaoChu commented on paddleseg/models/losses/focal_loss.py:
```diff
@@ -0,0 +1,54 @@
+        self.gamma = gamma
+        self.ignore_index = ignore_index
+
+    def forward(self, inp, target):
```
inp, target改成logit，label吧

LutaoChu commented on paddleseg/models/backbones/mobilenetv2.py:
```diff
@@ -0,0 +1,168 @@
+        The MobileNetV2 implementation based on PaddlePaddle.
+
+        The original article refers to
+        Song-Hai Zhanga, Xin Donga, Jia Lib, Ruilong Lia, Yong-Liang Yangc
```
Reference改成mobilenetv2

LutaoChu commented on paddleseg/models/losses/focal_loss.py:
```diff
@@ -0,0 +1,54 @@
+class FocalLoss(nn.Layer):
+    """
+    The Focal Loss implement of Portrait Net.
+
```
改成
'''
Focal Loss.

Code referenced from: 
https://github.com/clcarwin/focal_loss_pytorch/blob/master/focalloss.py
'''

LutaoChu commented on paddleseg/models/losses/focal_loss.py:
```diff
@@ -0,0 +1,54 @@
+        logpt = F.log_softmax(inp)
+        logpt = paddle.gather_nd(logpt, target)
+
+        pt = paddle.to_tensor(np.exp(logpt.numpy()))
```
为什么不用paddle.exp呢？

LutaoChu commented on paddleseg/models/losses/focal_loss.py:
```diff
@@ -0,0 +1,54 @@
+    def forward(self, inp, target):
+        inp = paddle.reshape(inp, [inp.shape[0], inp.shape[1], -1])
+        inp = paddle.transpose(inp, [0, 2, 1])
+        inp = paddle.reshape(inp, [-1, 2])
```
应该是inp = paddle.reshape(inp, [-1, inp.shape[2])吧，用的时候不一定是二分类

LutaoChu replied:
加上注释  ```# N,H*W,C => N*H*W,C```

LutaoChu commented on paddleseg/models/losses/focal_loss.py:
```diff
@@ -0,0 +1,54 @@
+        self.ignore_index = ignore_index
+
+    def forward(self, inp, target):
+        inp = paddle.reshape(inp, [inp.shape[0], inp.shape[1], -1])
```
加上注释   ```# N,C,H,W => N,C,H*W```

LutaoChu commented on paddleseg/models/losses/focal_loss.py:
```diff
@@ -0,0 +1,54 @@
+
+    def forward(self, inp, target):
+        inp = paddle.reshape(inp, [inp.shape[0], inp.shape[1], -1])
+        inp = paddle.transpose(inp, [0, 2, 1])
```
加上注释 ```# N,C,H*W => N,H*W,C```

LutaoChu commented on paddleseg/models/losses/kl_loss.py:
```diff
@@ -0,0 +1,41 @@
+        self.kl_loss = nn.KLDivLoss(reduction="mean")
+        self.temperature = temperature
+
+    def forward(self, inp, target):
```
inp, target改成logit，label

LutaoChu commented on paddleseg/models/losses/kl_loss.py:
```diff
@@ -0,0 +1,41 @@
+@manager.LOSSES.add_component
+class KLLoss(nn.Layer):
+    """
+    The Kullback-Leibler divergence Loss implement of Portrait Net.
```
去掉portraitnet
加上
```
    Code referenced from: 
    https://github.com/peterliht/knowledge-distillation-pytorch/blob/master/model/net.py
    
    Compute the knowledge-distillation (KD) loss given outputs, labels.
    "Hyperparameters": temperature and alpha
```

LutaoChu commented on paddleseg/models/losses/focal_loss.py:
```diff
@@ -0,0 +1,54 @@
+            and does not contribute to the input gradient. Default ``255``.
+    """
+
+    def __init__(self, gamma=2.0, ignore_index=255):
```
为了能让focal loss更通用，加上edge_label=False参数，通过edge_label判断是不是用edge监督

LutaoChu commented on paddleseg/models/losses/focal_loss.py:
```diff
@@ -0,0 +1,54 @@
+    def __init__(self, gamma=2.0, ignore_index=255):
+        super(FocalLoss, self).__init__()
+        self.gamma = gamma
+        self.ignore_index = ignore_index
```
加上self.edge_label = edge_label

LutaoChu commented on paddleseg/core/train.py:
```diff
@@ -42,6 +42,11 @@ def loss_computation(logits_list, labels, losses, edges=None):
         # Whether to use edges as labels According to loss type.
         if loss_i.__class__.__name__ in ('BCELoss', ) and loss_i.edge_label:
```
改成 if loss_i.__class__.__name__ in ('BCELoss',  "FocalLoss") and loss_i.edge_label:

LutaoChu commented on paddleseg/core/train.py:
```diff
@@ -42,6 +42,11 @@ def loss_computation(logits_list, labels, losses, edges=None):
         if loss_i.__class__.__name__ in ('BCELoss', ) and loss_i.edge_label:
             loss_list.append(losses['coef'][i] * loss_i(logits, edges))
+        elif loss_i.__class__.__name__ in ("FocalLoss", ):
+            loss_list.append(losses['coef'][i] * loss_i(logits, edges))
```
去掉上面两行

txyugood committed:
1.Modify the reference of mobilenet v2, kl loss and focal loss.
2.Add common transform for inputs of portrait net.
3.Standardize some variable names.

LutaoChu committed:
Update README.md

LutaoChu commented on paddleseg/models/portraitnet.py:
```diff
@@ -0,0 +1,215 @@
+
+
+class PortraitNetHead(nn.Layer):
+    def __init__(self, num_classes, add_edge=False):
```
开放参数min_channel，channel_ratio吧

LutaoChu commented on paddleseg/models/portraitnet.py:
```diff
@@ -0,0 +1,215 @@
+        pretrained (str, optional): The path or url of pretrained model. Default: None
+    """
+
+    def __init__(self, num_classes, backbone, add_edge=False, pretrained=None):
```
开放参数min_channel，channel_ratio吧

LutaoChu commented on paddleseg/models/backbones/mobilenetv2.py:
```diff
@@ -0,0 +1,168 @@
+
+
+@manager.MODELS.add_component
+class MobileNetV2(nn.Layer):
```
提到文件最前头吧

LutaoChu committed:
Update portraitnet_eg1800_224x224_46k.yml

LutaoChu committed:
Update supervisely.py

LutaoChu committed:
Update eg1800.py

LutaoChu committed:
Update README.md

txyugood committed:
Update portraitnet.py and mobilenetv2.py.

LutaoChu committed:
Merge branch 'develop' into dev
